In [1]:
import numpy as np
import seaborn as sns
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import OneHotEncoder
%matplotlib inline

In [2]:
df_train = pd.read_csv(r"C:\Users\praka\Desktop\Studies\AmEXkaggle\amex-default-prediction\train_data.csv")

In [3]:
df_head = df_train.head()

In [4]:
df_head.to_csv('train_head.csv')

In [5]:
df_train.tail()

customer_ID         S_2  \
5531446  fffff1d38b785cef84adeace64f8f83db3a0c31e8d92ea...  2017-11-05   
5531447  fffff1d38b785cef84adeace64f8f83db3a0c31e8d92ea...  2017-12-23   
5531448  fffff1d38b785cef84adeace64f8f83db3a0c31e8d92ea...  2018-01-06   
5531449  fffff1d38b785cef84adeace64f8f83db3a0c31e8d92ea...  2018-02-06   
5531450  fffff1d38b785cef84adeace64f8f83db3a0c31e8d92ea...  2018-03-14   

              P_2      D_39       B_1       B_2       R_1       S_3      D_41  \
5531446  0.979332  0.416013  0.020818  0.828199  0.003487  0.090743  0.005340   
5531447  0.984907  0.296536  0.007209  0.812610  0.005904  0.079886  0.002243   
5531448  0.983019  0.443984  0.013151  0.815422  0.003457  0.100503  0.002111   
5531449  0.969861  0.442553  0.009855  1.003541  0.005117  0.101802  0.009930   
5531450  0.982175  0.002474  0.000077  0.992880  0.000809  0.119165  0.003287   

              B_3  ...  D_136  D_137  D_138     D_139     D_140     D_141  \
5531446  0.025139  ...    NaN    NaN    NaN  0.006838  0.003680  0.000457   
5531447  0.023691  ...    NaN    NaN    NaN  0.003310  0.007097  0.007857   
5531448  0.012343  ...    NaN    NaN    NaN  0.009955  0.009994  0.001088   
5531449  0.008578  ...    NaN    NaN    NaN  0.005541  0.006564  0.009883   
5531450  0.014092  ...    NaN    NaN    NaN  0.007316  0.002888  0.006207   

         D_142     D_143     D_144     D_145  
5531446    NaN  0.000905  0.001498  0.002774  
5531447    NaN  0.002777  0.008225  0.008856  
5531448    NaN  0.005693  0.006773  0.005566  
5531449    NaN  0.008123  0.001168  0.003983  
5531450    NaN  0.005110  0.003184  0.001914  

[5 rows x 190 columns]

In [6]:
df_train.shape

(5531451, 190)

In [7]:
for col in df_train.columns:
    print(f"{col} \t {df_train[col].dtype}")

customer_ID 	 object
S_2 	 object
P_2 	 float64
D_39 	 float64
B_1 	 float64
B_2 	 float64
R_1 	 float64
S_3 	 float64
D_41 	 float64
B_3 	 float64
D_42 	 float64
D_43 	 float64
D_44 	 float64
B_4 	 float64
D_45 	 float64
B_5 	 float64
R_2 	 float64
D_46 	 float64
D_47 	 float64
D_48 	 float64
D_49 	 float64
B_6 	 float64
B_7 	 float64
B_8 	 float64
D_50 	 float64
D_51 	 float64
B_9 	 float64
R_3 	 float64
D_52 	 float64
P_3 	 float64
B_10 	 float64
D_53 	 float64
S_5 	 float64
B_11 	 float64
S_6 	 float64
D_54 	 float64
R_4 	 float64
S_7 	 float64
B_12 	 float64
S_8 	 float64
D_55 	 float64
D_56 	 float64
B_13 	 float64
R_5 	 float64
D_58 	 float64
S_9 	 float64
B_14 	 float64
D_59 	 float64
D_60 	 float64
D_61 	 float64
B_15 	 float64
S_11 	 float64
D_62 	 float64
D_63 	 object
D_64 	 object
D_65 	 float64
B_16 	 float64
B_17 	 float64
B_18 	 float64
B_19 	 float64
D_66 	 float64
B_20 	 float64
D_68 	 float64
S_12 	 float64
R_6 	 float64
S_13 	 float64
B_21 	 float64
D_69 	 float64
B

In [8]:
df_train['D_63']

0          CR
1          CR
2          CR
3          CR
4          CR
           ..
5531446    CL
5531447    CL
5531448    CL
5531449    CL
5531450    CL
Name: D_63, Length: 5531451, dtype: object

In [9]:
df_train['D_63'].value_counts()

CO    4119621
CR     930133
CL     438390
XZ      25786
XM      10556
XL       6965
Name: D_63, dtype: int64

In [10]:
df_train['D_64']

0          O
1          O
2          O
3          O
4          O
          ..
5531446    O
5531447    O
5531448    O
5531449    O
5531450    O
Name: D_64, Length: 5531451, dtype: object

In [11]:
df_train['D_64'].value_counts()

O     2913244
U     1523448
R      840112
-1      37205
Name: D_64, dtype: int64

In [12]:
df_train['D_63'] = df_train['D_63'].astype('category')
df_train['D_64'] = df_train['D_64'].astype('category')

In [13]:
df_train['D_63_new'] = df_train['D_63'].cat.codes
df_train['D_64_new'] = df_train['D_64'].cat.codes

In [14]:
enc=OneHotEncoder()
enc_data=pd.DataFrame(enc.fit_transform(df_train[['D_63_new','D_64_new']]).toarray())

In [15]:
df_train = df_train.join(enc_data)

In [16]:
df_train['S_2'] = pd.to_datetime(df_train['S_2'])

In [17]:
df_train.dtypes

customer_ID            object
S_2            datetime64[ns]
P_2                   float64
D_39                  float64
B_1                   float64
                    ...      
6                     float64
7                     float64
8                     float64
9                     float64
10                    float64
Length: 203, dtype: object

In [18]:
df_train.count()

customer_ID    5531451
S_2            5531451
P_2            5485466
D_39           5531451
B_1            5531451
                ...   
6              5531451
7              5531451
8              5531451
9              5531451
10             5531451
Length: 203, dtype: int64

In [19]:
df_train.describe(include='all')

<ipython-input-19-b75fe9c527f9>:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  df_train.describe(include='all')


customer_ID  \
count                                             5531451   
unique                                             458913   
top     08c57c902d75c37c27c892b986b16942337d2a0659a0e2...   
freq                                                   13   
first                                                 NaN   
last                                                  NaN   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                        S_2           P_2          D_39           B_1  \
count               5531451  5.485466e+06  5.531451e+06  5.531451e+06   
unique                  396           NaN           NaN           NaN   
top     2018-02-17 00:00:00           NaN           NaN           NaN   
freq                  26715           NaN           NaN           NaN   
first   2017-03-01 00:00:00           NaN           NaN           NaN   
last    2018-03-31 00:00:00           NaN           NaN           NaN   
mean                    NaN  6.563340e-01  1.531172e-01  1.240100e-01   
std                     NaN  2.446494e-01  2.700709e-01  2.119869e-01   
min                     NaN -4.589548e-01  5.026190e-09 -7.588799e+00   
25%                     NaN  4.803307e-01  4.528464e-03  8.863645e-03   
50%                     NaN  6.942950e-01  9.056902e-03  3.132968e-02   
75%                     NaN  8.648159e-01  2.366407e-01  1.259019e-01   
max                     NaN  1.010000e+00  5.389619e+00  1.324060e+00   

                 B_2           R_1           S_3          D_41           B_3  \
count   5.529435e+06  5.531451e+06  4.510907e+06  5.529435e+06  5.529435e+06   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
first            NaN           NaN           NaN           NaN           NaN   
last             NaN           NaN           NaN           NaN           NaN   
mean    6.214887e-01  7.880270e-02  2.258455e-01  5.978469e-02  1.325389e-01   
std     4.014877e-01  2.263971e-01  1.933475e-01  2.025443e-01  2.349929e-01   
min     9.192280e-09  1.534223e-09 -6.271320e-01  5.566545e-10  6.285293e-09   
25%     1.053313e-01  2.895934e-03  1.272588e-01  2.873244e-03  5.227570e-03   
50%     8.143328e-01  5.782230e-03  1.639082e-01  5.746725e-03  9.777230e-03   
75%     1.002403e+00  8.660590e-03  2.581017e-01  8.615665e-03  1.550507e-01   
max     1.010000e+00  3.256284e+00  5.482888e+00  8.988807e+00  1.625262e+00   

        ...             1             2             3             4  \
count   ...  5.531451e+06  5.531451e+06  5.531451e+06  5.531451e+06   
unique  ...           NaN           NaN           NaN           NaN   
top     ...           NaN           NaN           NaN           NaN   
freq    ...           NaN           NaN           NaN           NaN   
first   ...           NaN           NaN           NaN           NaN   
last    ...           NaN           NaN           NaN           NaN   
mean    ...  7.447632e-01  1.681535e-01  1.259163e-03  1.908360e-03   
std     ...  4.359943e-01  3.740026e-01  3.546235e-02  4.364308e-02   
min     ...  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%     ...  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%     ...  1.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
75%     ...  1.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
max     ...  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   

                   5             6             7    

In [20]:
df_na=pd.DataFrame(df_train.isna().sum())

In [21]:
df_na.head()

0
customer_ID      0
S_2              0
P_2          45985
D_39             0
B_1              0

In [28]:
for item in df_na['0']:
    print(item)

KeyError: '0'

In [22]:
display(df_na.to_string)

<bound method DataFrame.to_string of                  0
customer_ID      0
S_2              0
P_2          45985
D_39             0
B_1              0
...            ...
6                0
7                0
8                0
9                0
10               0

[203 rows x 1 columns]>

In [23]:
df_customer = df_train.groupby('customer_ID')

In [24]:
df_customer.head()

MemoryError: Unable to allocate 8.08 GiB for an array with shape (196, 5531451) and data type float64

In [ ]:
transaction_count = df_customer['S_2'].count()

In [ ]:
transaction_count.head()

In [ ]:
transaction_count.shape

In [ ]:
transaction_count.describe()

OBSERVATIONS:
    Unique Customers - 458913
    No. of transaction data/customer - 1 to 13
    

In [ ]:
df_train.isna().sum()/df_train.shape[0]

In [ ]:
# Reading the train labels
train_labels = pd.read_csv(r"C:\Users\praka\Desktop\Studies\AmEXkaggle\amex-default-prediction\train_labels.csv")

In [ ]:
train_labels.describe()

In [ ]:
train_labels.head()